### PR TITLE
hiredis: updated package definition to use CMake

### PR DESCRIPTION
**Pull request by lpottier**
Hiredis was using Make prior to 1.0.0 and then moved to CMake. I have updated the package to use Cmake and added variants for SSL.

Signed-off-by: Loïc Pottier <pottier1@llnl.gov>

### CODE REVIEW

**Comment by haampie** [comment]
Can you please use

https://spack.readthedocs.io/en/latest/packaging_guide.html#multiple-build-systems

There is not need to deprecate it over a build system change, and since `ccache` uses hiredis and should be easy to bootstrap, it would be a shame if it required building cmake and its dependencies first.

**Review by haampie — CHANGES_REQUESTED** [review]
Can you please use separate classes for the 2 build systems?

https://spack.readthedocs.io/en/latest/packaging_guide.html#multiple-build-systems

Deprecation shouldn't happen over build system changes, just over (sever) bugs that won't be fixed.

File: var/spack/repos/builtin/packages/hiredis/package.py
Comment (haampie) [inline var/spack/repos/builtin/packages/hiredis/package.py]:
```suggestion
    depends_on("cmake@3.18:", type="build", when="build_system=cmake")
```

File: var/spack/repos/builtin/packages/hiredis/package.py
Comment (haampie) [inline var/spack/repos/builtin/packages/hiredis/package.py:54]: Can you move the old Makefile related methods into

```python
class MakefileBuilder(spack.build_systems.makefile.MakefileBuilder):
    def install(...):
```

and

```python
class CMakeBuilder(spack.build_systems.cmake.CMakeBuilder):
    def cmake_args(...):
```

**Review by haampie — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/hiredis/package.py
Comment (lpottier) [reply var/spack/repos/builtin/packages/hiredis/package.py:54]: Yes, thanks for the tips. I just did it and I also removed the deprecated flags!

File: var/spack/repos/builtin/packages/hiredis/package.py
Comment (rblake-llnl) [inline var/spack/repos/builtin/packages/hiredis/package.py]: You can be sole maintainer, or you can add me to help share the load.  I think you're more familiar with the package.

File: var/spack/repos/builtin/packages/hiredis/package.py
Comment (rblake-llnl) [inline var/spack/repos/builtin/packages/hiredis/package.py:27]: I think `1:` and `:0` work better, right?  That way if 0.14.2 comes out it can be covered.

File: var/spack/repos/builtin/packages/hiredis/package.py
Comment (rblake-llnl) [inline var/spack/repos/builtin/packages/hiredis/package.py:61]: Does cmake also need the darwin fix?  If not, let's use your version.

**Review by rblake-llnl — COMMENTED** [review]
LGTM modulo some minor things.  @lpottier actually both had the same idea at the same time.

File: var/spack/repos/builtin/packages/hiredis/package.py
Comment (haampie) [reply var/spack/repos/builtin/packages/hiredis/package.py:61]: I believe CMake does rpaths better

File: var/spack/repos/builtin/packages/hiredis/package.py
Comment (lpottier) [reply var/spack/repos/builtin/packages/hiredis/package.py]: I am good with sharing the load! I will add you.

File: var/spack/repos/builtin/packages/hiredis/package.py
Comment (lpottier) [reply var/spack/repos/builtin/packages/hiredis/package.py:27]: Yes, good point! I have switched to `1:` and `:0`.

**Comment by bernhardkaindl** [comment]
@lpottier Are your updates to apply the requested changes finished?

> Can you please use separate classes for the 2 build systems?
> https://spack.readthedocs.io/en/latest/packaging_guide.html#multiple-build-systems
> Deprecation shouldn't happen over build system changes, just over (sever) bugs that won't be fixed.

On the 1st glance, I't looks it's not split into separate classes yet.
If you want another review, request it from haampie using the circular icon in his entry in the reviewers list on the top-right edge.

**Comment by lpottier** [comment]
> @lpottier Are your updates to apply the requested changes finished?
> 
> > Can you please use separate classes for the 2 build systems?
> > https://spack.readthedocs.io/en/latest/packaging_guide.html#multiple-build-systems
> > Deprecation shouldn't happen over build system changes, just over (sever) bugs that won't be fixed.
> 
> On the 1st glance, I't looks it's not split into separate classes yet. If you want another review, request it from haampie using the circular icon in his entry in the reviewers list on the top-right edge.

@bernhardkaindl I did split it into two classes (following [Spack multiple-build-systems](https://spack.readthedocs.io/en/latest/packaging_guide.html#multiple-build-systems)) . Unless, I am missing something here?

Otherwise, my updates are finished (unless the splitting into two classes is not correct per Spack guidelines)

**Review by haampie — APPROVED** [review]

**Comment by bernhardkaindl** [comment]
I only had a first glance, so thanks for confirming that your changes are finished.

**Pull request merged**
